lib/exla/shape.ex

```diff
@@ -0,0 +1,14 @@
+defmodule Exla.Shape do
+  alias __MODULE__, as: Shape
+
+  @enforce_keys [:dims, :dtype, :ref]
+  defstruct [:dims, :dtype, :ref]
+
+  # TODO: convert dtype to integer representation to use in NIF
+  def make_shape(dtype, dims) when is_atom(dtype) and is_tuple(dims) do
+    case Exla.NIF.make_shape(dtype, dims) do
+      {:ok, ref} -> {:ok, %Shape{ref: ref, dims: dims, dtype: dtype}}
+      {:error, msg} -> {:error, msg}
```

Comment:
Btw, do you have any ideas which kind of errors we may have here?

Reply:
As I've been debugging, I've been changing some of the argument checks within the NIF to output which argument specifically it doesn't like. It's much more informative then just throwing a badarg on that end. Once everything is cleaned up, it should be basically impossible to get an error from the NIF because everything should be thrown from the Elixir side. For development though it makes it nice to know where something is going wrong

Reply:
@seanmor5 if that's the case, we can probably rewrite it to this:

```elixir
      {:ok, ref} = Exla.NIF.make_shape(dtype, dims) do
      %Shape{ref: ref, dims: dims, dtype: dtype}
```

We still get an error but without pushing bureaucracy up the stack. WDYT?

Reply:
Okay I see, so it throws an error on the bad match and we still get to see what the problem is clearly but there's no need to keep passing that error along. This is the case for pretty much every NIF so I'll go back and edit them to match this